r/R/util.R

```diff
@@ -251,3 +251,25 @@ augment_io_error_msg <- function(e, call, schema = NULL, format = NULL) {
   handle_augmented_field_misuse(msg, call)
   abort(msg, call = call)
 }
+
+simulate_data_frame <- function(schema) {
```

Comment:
Since this is going to be called every time someone does `select/rename/relocate`, I'd like for this function to be cheaper. In other PRs I've been noticing the overhead of creating R6 objects, which generally is not terrible (~150 microseconds on my machine) but it adds up. And here, we're creating lots of objects we're throwing away: for each column, we create a Field, then a DataType from that, then in concat_arrays, we create a null DataType, an Array with that, and a new Array that is cast to the correct DataType. That adds up to around 1ms per column, every time this function is called. That's enough to get noticed.

Can we move this to C++? Should be a simple enough switch statement to map Arrow type ids to the corresponding R length-0 vector.

Reply:
Just to confirm, I benchmarked this function on the schema of the taxi dataset (20 columns), and the median time was 15ms, so a little under 1ms per column.

Reply:
Thanks for confirming! Currently working on a C++ simplification though may ask for help to get it over the line ahead of the release if I can't figure it out by the end of tomorrow.